### ISSUE TITLE
AssertionError and inconsistent sizes

### ISSUE THREAD

**Issue by reintroducing** [twolfson/gulp.spritesmith#57]
I just ran into this:

```
assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Normal sprite has inconsistent size with retina sprite. "app-store-small" is 122x36 while "app-store-small@2x" is 243x72.
    at getRetinaGroups (/Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/lib/gulp-spritesmith.js:288:11)
    at Array.map (native)
    at handleSpritesheets (/Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/lib/gulp-spritesmith.js:285:36)
    at /Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/async/lib/async.js:251:17
    at done (/Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/async/lib/async.js:132:19)
    at /Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/async/lib/async.js:32:16
    at /Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/async/lib/async.js:248:21
    at /Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/async/lib/async.js:572:34
    at smithCallbackData (/Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/spritesmith/src/smith.js:172:7)
    at fn (/Users/mprzybylski/Desktop/threshold-test/node_modules/gulp.spritesmith/node_modules/spritesmith/node_modules/async/lib/async.js:582:34)
```

While I understand why this is happening, I think there should be some leeway here in terms of 1 pixel off due to odd number dimensions. 

Allow me to explain: I am making my sprites sized perfectly as they should be based on designs. Sometimes that means the 2x versions will end in odd numbers like the above example. I then use gulp-image-resize to create 1x versions which tends to round the numbers for obvious reasons (even if I didn't use this, they'd get rounded one way or another). Then the finalized images are run through spritesmith during a gulp build.

I know the easy answer is "just make sure you're creating images with even numbered dimension ends" but I don't think that should be a limitation.  I have been using this plugin successfully to do this for a while now without using the built in retina functionality (I'd be happy to share how if you'd like) but am trying to consolidate the task to be used the way its meant to be used.  I wish it allowed to be off one pixel without throwing the error above.

Thoughts?

**Comment by twolfson** [comment]
I believe there is some confusion over how retina sprites operate. The goal is to achieve a higher pixel density than is possible with normal images. We are creating an image that is 2x as large as a normal image but then scaled down by 2x. This leads to defining more interim pixel values. For example, `x=1` and `x=2` are defined for normal ones but `x=1.5` is interpolated. When we perform scaling for retina images though, `x=2`, `x=3`, and `x=4` are defined which translate to `x=1`, `x=1.5`, and `x=2` when scaled down.

For your case, if we were to remove sprites/spritesheets from the equation and use the source images (e.g. `app-store-small.png`), then we would be defining non-integer widths/heights (e.g. `width=1.5`). This would lead to the browser scaling our image and likely causing some bad scaling.

To solve this, typically people will pad their images with transparent padding until they reach an integer width/height. Now that we know the solution for the normal image case, we would translate this to retina images by adding 2x the amount of transparent padding to our 2x scaled up image.

tl;dr - Retina sprites should be the scaled up version of normal sprites, not vice versa. Adding transparent padding to your images will solve the assertion error as well as scaling issues for the normal sprites.

**Comment by reintroducing** [comment]
Hey @twolfson,
I don't mean to beat a dead horse but I don't really know anyone that would ever scale up to create a retina sprite because that would cause artifacting. I've always gotten retina sized PSDs from designers and saved out the retina version then (if generating manually) gone in and image resized through PS to 50%/50% and that is how you'd end up getting non-whole numbers which subsequently get rounded up or down hence the difference.  

Of course this is easily fixed as you said by just adding an extra integer to whichever dimension needs it when creating the retina version (of dead pixels) and would require remembering to do this every time (which could be error prone).  

I don't know all the internals of the plugin itself and if there is a possibility for a solution that would provide a happy medium but regardless I want you to know that the plugin is fantastic and I will continue to use it regardless.  My current setup that creates retina sprites separately (not the "plugin" way) works as is so I can just stick to the unofficial way of doing things.

Keep up the great work.

**Comment by twolfson** [comment]
Ah, sorry if I was unclear. I was talking partially in theory and partially in application. When I mean "transferring", I meant the numerical values, not actually upscaling an image as that would defeat the purpose of retina altogether =P

Unfortunately, we need to retain current system (where we perform asserts) as it serves 2 purposes:

- Catch mismatched sprites (e.g. if we apply a retina src filter but the ordering of normal and retina sprites is incorrect)
- Inform persons who were given bad non-2x sprites from an external source (rather than making their own) about the invalidity of those sprites (e.g. if there is a 10px wide 1x sprite with a 25px wide 2x sprite -- this is something we should inform them about)

**Comment by nlound** [comment]
I fully understand how retina works but: Is there any way to bypass the size check?

**Comment by twolfson** [comment]
@nlound Can you explain your use case further?

**Comment by nlound** [comment]
There is a new feature in Adobe Photoshop CC for exporting assets in several sizes with a single action. It doesn't work properly but my team want to use it anyway (dont ask why).

Retina images are not resized properly and one or two pixels are gone when exporting non rasterized layers. (more information [here](https://forums.adobe.com/message/8551963))

I am actually working in a plugin to preprocess images before using spritesmith for adding transparent pixels if the retina image is not retina.

**Comment by twolfson** [comment]
Ah, k. To answer your initial question, the current implementation doesn't support ignoring mismatched retina sizes; we stop the streams at that point.

https://github.com/twolfson/gulp.spritesmith/blob/6.2.0/lib/gulp-spritesmith.js#L203-L226

As for the plugin/task, I suggest using something like `gulp-gm`:

https://github.com/scalableminds/gulp-gm/tree/v0.0.8#async-manipulation

```js
gulp.src('*@2x.png')
  .pipe(gm(function handleGm (gmfile, done) {
    // Retrieve the current image size
    gmfile.size(function handleSize (err, size) {
      // If there was an error, callback with it
      if (err) { return done(err); }

      // Otherwise, if the size is even, then do nothing
      if (size.width % 2 === 0 && size.height % 2 === 0) {
        return done(null, gmfile);
      }

      // Otherwise, extend our image to even dimensions
      // http://www.imagemagick.org/Usage/crop/#extent
      return done(null, gmfile
        .background('transparent')
        .gravity('northwest')
        .extent(size.width + (size.width % 2), size.height + (size.height % 2)));
    });
  }))
  .pipe(gulp.dest('src'));
```

**Comment by nlound** [comment]
Thank you very much sir!

**Comment by NickVanHolsbeeck** [comment]
Is this fixed yet so we could ignore mismatched retina sizes? Or should we still go with gulp-gm? (This issue is from march 2016, so ...)

**Comment by twolfson** [comment]
@NickVanHolsbeeck Everyone's original questions/concerns were addressed and this issue is considered resolved. If you have additional questions outside of this thread, please open a separate issue to avoid spamming the existing subscribers
